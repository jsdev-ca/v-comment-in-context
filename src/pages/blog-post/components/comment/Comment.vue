<template>
  <!--
    The opening `div` shouldn’t be an `article`, as there’s a chance that the
    comments won’t make much (if any) sense outside of their context:

    https://html.spec.whatwg.org/multipage/sections.html#the-article-element

    Should it be a `section`, then? It’s unclear that it should be, as it isn’t
    obvious that it’s a sectioning content element.
  -->
  <div class="container">
    <div :style="{ margin: '0 10px' }">
      <!-- Leave `data-testid` alone; our unit test needs it.  -->
      <avatar data-testid="avatar" :alt="alt" :url="url" />
    </div>
    <div :style="{ margin: '0 10px' }">
      <div class="metadata">
        <div class="bold">
          {{ author }}
        </div>
        <badge :styles="colors" :text="contributorType" />
        <div v-if="whenCommentWasPosted" class="converted-timestamp"><!--
          This comment exists to ensure there’s no additional space between
          the “•” and the value.
          -->{{ whenCommentWasPosted }}
        </div>
      </div>
      <p class="comments" v-for="(text, i) in comments" :key="i">
        {{ text }}
      </p>
      <div class="calls-to-action">
        <button
          v-on:click="showCommentCard"
          :aria-label="`respond to ${author}’s comment`"
        >
          Reply
        </button>
        <button
          v-on:click="showReplies"
          :aria-label="`read the ${
            noun === 'reply' ? 'lone' : numReplies
          } ${noun} to ${author}’s comment`"
          :disabled="numReplies === 0"
          :style="{ cursor: numReplies === 0 ? 'not-allowed' : 'pointer'}"
        >
          <span :style="{ color: 'var(--grey)', marginRight: '5px' }">
            {{ numReplies }}
          </span>
          {{ noun }}
        </button>
        <div class="upvotes">
          <button
            v-on:click="dynamicUpvotes += 1"
            :aria-label="`upvote ${author}’s comment`"
          >
            <i class="gg-chevron-up" />
          </button>
          {{ dynamicUpvotes }}
        </div>
        <div class="downvotes">
          <button
            v-on:click="dynamicDownvotes += 1"
            :aria-label="`downvote ${author}’s comment`"
          >
            <i class="gg-chevron-down" />
          </button>
          {{ dynamicDownvotes }}
        </div>
      </div>
    </div>
  </div>
</template>

<script>
function showCommentCard() {}

function showReplies() {}

export { showCommentCard, showReplies };

export default {
  data() {
    return {
      colors: {
        author: {
          background: '#333',
          color: '#fff'
        },
        guest: {
          background: '#0052cc',
          color: '#fff'
        }
      }[this.contributorType],
      dynamicDownvotes: this.downvotes,
      dynamicUpvotes: this.upvotes
    };
  },
  computed: {
    alt: function () {
      return `The avatar of the commenter, ${this.author}.`;
    },
    noun: function () {
      return this.numReplies === 1 ? 'reply' : 'replies';
    },
    url: function () {
      return /^.+\.(png|jpe?g|gif)$/i.test(this.avatarFilename)
        ? `../../../static/${this.avatarFilename}`
        : null;
    }
  },
  methods: {
    showCommentCard,
    showReplies
  },
  // N.B. Vue won’t provide the specified default for `null` values, so avoid
  // using `null` in the database as much as possible.

  // On a different vein, would need to speak with the designers and managers
  // about the fallback options, as well as the point at which it wouldn’t be
  // sensible to render the component at all.
  props: {
    author: {
      type: String,
      required: true,
      validator: string => string !== ''
    },
    avatarFilename: String,
    comments: {
      type: Array,
      required: true,
      validator: array => array.length > 0 && array.every(item => (
        typeof item === 'string' && item !== ''
      ))
    },
    contributorType: {
      type: String,
      required: true,
      validator: string => ['author', 'guest'].includes(string)
    },
    downvotes: {
      type: Number,
      required: true
    },
    numReplies: {
      type: Number,
      required: true
    },
    upvotes: {
      type: Number,
      required: true
    },
    whenCommentWasPosted: String
  }
};
</script>

<style lang="scss" scoped>
// Would make this a global mixin if it made sense to do so.
@mixin display-text-small($color) {
  color: $color;
  font-size: 1.2rem;
  text-transform: uppercase;
}

.container {
  border: 1px solid #ebebeb;
  border-left: 0;
  border-right: 0;
  display: flex;
  margin: 20px auto;
  min-height: 110px;
  padding: 20px 0;
  width: 600px;

  &:hover {
    .calls-to-action,
    button {
      color: var(--grey);
    }
  }
}

.metadata {
  align-items: center;
  display: flex;

  & > * {
    margin-right: 5px;

    &.converted-timestamp {
      @include display-text-small(var(--grey));

      &::before {
        content: "•";
        margin-right: 5px;
      }
    }
  }
}

.comments {
  margin: 12px 0;
  // Q: Why `80ch`?
  // A: https://www.w3.org/WAI/tutorials/page-structure/styling/#line-length
  max-width: 80ch;
}

.calls-to-action {
  align-items: center;
  color: var(--silver);
  display: flex;
  font-size: 1.2rem;

  & > * {
    &:not(:last-child) {
      margin-right: 16px;
    }
  }

  [class$="votes"] {
    align-items: center;
    display: inline-flex;
  }

  button {
    @include display-text-small(var(--silver));

    background: none;
    border: 0;
    display: inline-flex;
    padding: 0;

    [class^="gg-chevron-"] {
      --ggs: .8;
    }
  }
}
</style>
