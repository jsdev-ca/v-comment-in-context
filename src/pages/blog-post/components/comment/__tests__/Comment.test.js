import { render, fireEvent } from '@testing-library/vue';
import { config, mount } from '@vue/test-utils';

import Comment, { showCommentCard, showReplies } from '../Comment.vue';

config.showDeprecationWarnings = false;

describe('Comment', () => {
  const options = {
    methods: {},
    propsData: {
      author: 'Yoda',
      avatarFilename: 'yoda.png',
      comments: [
        'Turned to the dark side, Anakin Skywalker has.'
      ],
      contributorType: 'author',
      downvotes: 1,
      numReplies: 100,
      upvotes: 1000
    },
    stubs: [
      'avatar',
      'badge'
    ]
  };
  let localOptions;

  beforeEach(() => {
    localOptions = JSON.parse(JSON.stringify(options));
  });

  it('matches the existing snapshot', () => {
    const wrapper = mount(Comment, localOptions);

    expect(wrapper.element).toMatchSnapshot();
  });

  describe('avatar', () => {
    it('has a `url` attribute with a valid value when the filename is valid', () => {
      /**
       * Alas, we can’t get our avatar stub via `getByAltText(
       *   /^The avatar of the commenter, .+\.$/
       * )`. Oh well.
       */
      const { getByTestId } = render(Comment, localOptions);
      const avatar = getByTestId('avatar');

      expect(avatar).toHaveAttribute(
        'url',
        `../../../static/${localOptions.propsData.avatarFilename}`
      );
    });

    it('lacks a `url` attribute when the filename is invalid', () => {
      localOptions.propsData.avatarFilename = 'yoda.jaypeg';

      const { getByTestId } = render(Comment, localOptions);
      const avatar = getByTestId('avatar');

      expect(avatar).not.toHaveAttribute('url');
    });
  });

  describe('reply button', () => {
    it('returns `undefined` when clicked', async () => {
      localOptions.methods.showCommentCard = jest.fn(showCommentCard);

      const { getByLabelText } = render(Comment, localOptions);
      const showCommentCardButton = getByLabelText(/^respond to .+’s comment$/);
      const spy = jest.spyOn(localOptions.methods, 'showCommentCard');

      await fireEvent.click(showCommentCardButton);

      expect(spy).toHaveReturnedWith(undefined);
    });
  });

  describe('{n} replies button', () => {
    describe('when there are 0 replies', () => {
      it('is disabled', () => {
        localOptions.propsData.numReplies = 0;

        const { getByLabelText } = render(Comment, localOptions);
        const showReplyButton = getByLabelText(/^read the 0 replies to .+’s comment$/);

        expect(showReplyButton).toBeDisabled();
      });
    });

    describe('when there’s 1 reply', () => {
      it('is enabled and has “read the lone reply to {authorName}’s comment” as its ARIA label', () => {
        localOptions.propsData.numReplies = 1;

        const { getByLabelText } = render(Comment, localOptions);
        const showReplyButton = getByLabelText(/^read the lone reply to .+’s comment$/);

        expect(showReplyButton).toBeEnabled();
      });
    });

    describe('when there are 100 replies', () => {
      it('is enabled and has “read the 100 replies to {authorName}’s comment” as its ARIA label', () => {
        const { getByLabelText } = render(Comment, localOptions);
        const showReplyButton = getByLabelText(/^read the 100 replies to .+’s comment$/);

        expect(showReplyButton).toBeEnabled();
      });

      it('returns `undefined` when clicked', async () => {
        localOptions.methods.showReplies = jest.fn(showReplies);

        const { getByLabelText } = render(Comment, localOptions);
        const showReplyButton = getByLabelText(/^read the 100 replies to .+’s comment$/);
        const spy = jest.spyOn(localOptions.methods, 'showReplies');

        await fireEvent.click(showReplyButton);

        expect(spy).toHaveReturnedWith(undefined);
      });
    });
  });

  describe('upvotes', () => {
    it('increment by 1 on upvote button click', async () => {
      const { getByLabelText, getByText } = render(Comment, localOptions);
      const upvoteButton = getByLabelText(/^upvote .+’s comment$/i);
  
      getByText('1000');

      await fireEvent.click(upvoteButton);
    
      getByText('1001');
    });
  });

  describe('downvotes', () => {
    it('increment by 1 on downvote button click', async () => {
      const { getByLabelText, getByText } = render(Comment, localOptions);
      const downvoteButton = getByLabelText(/^downvote .+’s comment$/i);

      getByText('1');

      await fireEvent.click(downvoteButton);

      getByText('2');
    });
  });
});
