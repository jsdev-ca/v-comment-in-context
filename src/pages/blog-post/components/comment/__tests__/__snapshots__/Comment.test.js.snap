// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment matches the existing snapshot 1`] = `
<div
  class="container"
>
  <div
    style="margin: 0px 10px;"
  >
    <avatar-stub
      alt="The avatar of the commenter, Yoda."
      data-testid="avatar"
      url="../../../static/yoda.png"
    />
  </div>
   
  <div
    style="margin: 0px 10px;"
  >
    <div
      class="metadata"
    >
      <div
        class="bold"
      >
        
        Yoda
      
      </div>
       
      <badge-stub
        styles="[object Object]"
        text="author"
      />
       
      <!---->
    </div>
     
    <p
      class="comments"
    >
      
      Turned to the dark side, Anakin Skywalker has.
    
    </p>
     
    <div
      class="calls-to-action"
    >
      <button
        aria-label="respond to Yoda’s comment"
      >
        
        Reply
      
      </button>
       
      <button
        aria-label="read the 100 replies to Yoda’s comment"
        style="cursor: pointer;"
      >
        <span
          style="margin-right: 5px;"
        >
          
          100
        
        </span>
        
        replies
      
      </button>
       
      <div
        class="upvotes"
      >
        <button
          aria-label="upvote Yoda’s comment"
        >
          <i
            class="gg-chevron-up"
          />
        </button>
        
        1000
      
      </div>
       
      <div
        class="downvotes"
      >
        <button
          aria-label="downvote Yoda’s comment"
        >
          <i
            class="gg-chevron-down"
          />
        </button>
        
        1
      
      </div>
    </div>
  </div>
</div>
`;
